File: Data Structures & Algorithms/search-2d-matrix/submission-2.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row = matrix.size(), col = matrix[0].size();
        int size = row * col;
        int l = 0, r = size - 1;

        while(l <= r) {
            int mid = l + (r-l)/2;
            int num = matrix[mid/col][mid%col];
            if(target == num) {
                return true;
            }
            else if(target < num) {
                r = mid - 1;
            }
            else l = mid + 1;
        }
        return false;
    }
};

File: Data Structures & Algorithms/valid-sudoku/submission-2.cs
public class Solution {
    public bool IsValidSudoku(char[][] board) {
        //key is index or row/row/block
        Dictionary<int, HashSet<char>> row = new();
        Dictionary<int, HashSet<char>> col = new();
        Dictionary<int, HashSet<char>> block = new();

         for(int i = 0; i < 9 ; i++) {
            for(int j = 0; j < 9; j ++) {
                if(board[i][j] == '.') {
                    continue;
                }
                int blockIdx = (i/3) * 3 + (j/3);
                if(!row.ContainsKey(i)) row[i] = new HashSet<char>();
                if(!col.ContainsKey(j)) col[j] = new HashSet<char>();
                if(!block.ContainsKey(blockIdx)) block[blockIdx] = new HashSet<char>();
                
                char num = board[i][j];
                if(row[i].Contains(num) || col[j].Contains(num) || block[blockIdx].Contains(num))
                {
                    return false;
                }
                row[i].Add(num);
                col[j].Add(num);
                block[blockIdx].Add(num);
            }
         }
         return true;
    }
}
